File: 프로그래머스/lv1/지준혁/덧칠하기.java
/*
1. 아이디어: 
2. 시간복잡도: O(n)
3. 자료구조:
*/

class Solution {
    public int solution(int n, int m, int[] section) {
        int pos = section[0] + m;
        int ans = 1;
        for (int i = 1; i < section.length; ++i) {
            if (pos > section[i]) {
                continue;
            }
            pos = section[i] + m;
            ++ans;
        }
        return ans;
    }
}

File: 프로그래머스/lv1/이하은/덧칠하기.kt
class Solution {
    val MAX = 100001
    var N = 0
    var M = 0
    
    val painted = BooleanArray(MAX) {true}
    var ans = Int.MAX_VALUE
    val temp = mutableListOf<Int>()
    
    fun solution(n: Int, m: Int, section: IntArray): Int {
        N = n
        M = m
        
        for(i in section){
            painted[i] = false
        }
        
        dfs(start = 1, cnt = 0)
        
        return ans
    }
    
    fun isAllPainted() = painted.all { it == true }
    
    fun dfs(start: Int, cnt: Int){        
        if(isAllPainted()){
            ans = minOf(ans, cnt)
            return
        }
        
        if(start > N - M + 1){
            if(isAllPainted()){
                ans = minOf(ans, cnt)
            }
            return
        }
        
        for(i in start until N - M + 2){
            if(!painted[i]){
                for(j in i until i + M){
                    if(!painted[j]) {
                        painted[j] = true
                        temp.add(j)
                    }
                }
                
                dfs(i + 1, cnt + 1)
                
                for(idx in temp){
                    painted[idx] = false
                }
                temp.clear()
            }
        }
    }
}